File: Sources/BraintreeCard/BTAuthenticationInsight.m
#import "BTAuthenticationInsight_Internal.h"

// Swift Module Imports
#if __has_include(<Braintree/Braintree-Swift.h>) // CocoaPods
#import <Braintree/Braintree-Swift.h>

#elif SWIFT_PACKAGE                              // SPM
/* Use @import for SPM support
 * See https://forums.swift.org/t/using-a-swift-package-in-a-mixed-swift-and-objective-c-project/27348
 */
@import BraintreeCore;

#elif __has_include("Braintree-Swift.h")         // CocoaPods for ReactNative
/* Use quoted style when importing Swift headers for ReactNative support
 * See https://github.com/braintree/braintree_ios/issues/671
 */
#import "Braintree-Swift.h"

#else                                            // Carthage
#import <BraintreeCore/BraintreeCore-Swift.h>
#endif

@implementation BTAuthenticationInsight

- (instancetype)initWithJSON:(BTJSON *)json {
    if (self = [super init]) {
        NSString *regulationEnvironment;
        
        if ([json[@"customerAuthenticationRegulationEnvironment"] asString]) {
            regulationEnvironment = [json[@"customerAuthenticationRegulationEnvironment"] asString];
        } else if ([json[@"regulationEnvironment"] asString]) {
            regulationEnvironment = [json[@"regulationEnvironment"] asString];
        }
        
        // GraphQL returns "PSDTWO" instead of "psd2"
        if ([regulationEnvironment isEqualToString:@"PSDTWO"]) {
            regulationEnvironment = @"psd2";
        }
        
        if (regulationEnvironment) {
            regulationEnvironment = regulationEnvironment.lowercaseString;
        }
        
        _regulationEnvironment = regulationEnvironment;
    }
    return self;
}

@end

File: Sources/BraintreeCard/BTAuthenticationInsight.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// Information pertaining to the regulatory environment for a credit card if authentication insight is requested during tokenization.
@objcMembers public class BTAuthenticationInsight: NSObject {

    // MARK: - Public Properties

    /// The regulation environment for the associated nonce to help determine the need for 3D Secure.
    /// See https://developer.paypal.com/braintree/docs/guides/3d-secure/advanced-options/ios/v5#authentication-insight
    /// for a list of possible values.
    public var regulationEnvironment: String?

    // MARK: - Initializer

    init(json: BTJSON) {
        if let customerAuthRegulationEnvironment = json["customerAuthenticationRegulationEnvironment"].asString() {
            self.regulationEnvironment = customerAuthRegulationEnvironment
        } else if let regulationEnvironment = json["regulationEnvironment"].asString() {
            self.regulationEnvironment = regulationEnvironment
        }

        // GraphQL returns "PSDTWO" instead of "psd2"
        if regulationEnvironment == "PSDTWO" {
            self.regulationEnvironment = "psd2"
        }

        if let regulationEnvironment {
            self.regulationEnvironment = regulationEnvironment.lowercased()
        }
    }
}
